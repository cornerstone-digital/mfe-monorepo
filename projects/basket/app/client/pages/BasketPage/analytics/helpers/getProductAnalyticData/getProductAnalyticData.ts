import getAllowanceByType from '@web-shop-core/helpers/getAllowanceByType'
import { isBingo } from '@helpers/typeCheck'

import { ProductAnalyticsData } from './getProductAnalyticsData.types'
import getTradeInParams from '../getTradeInParams'
// FIXME Refactor to handle individual object shapes to not require below hack
type AnalyticValues = BasketV2.Bundle & BasketV2.Hardware & BasketV2.Service & BasketV2.Discount

const getProductAnalyticData = (
  basket: BasketV2.Basket,
  specificPackage?: string,
  addonId?: string,
  hasAccessoryAddon?: boolean,
): object => {
  //* these definitions need to match the names from the UDL data layer */
  const allData: ProductAnalyticsData = {
    productPackageId: [],
    productBrand: [],
    productCapacity: [],
    productColour: [],
    productId: [],
    productModel: [],
    productName: [],
    productPlanData: [],
    productPackagePlanDuration: [],
    productPlanMinutes: [],
    productPlanSms: [],
    productPrice: [],
    productPriceNet: [],
    productPriceRecurrenceAmount: [],
    productPriceRecurrenceAmountNet: [],
    productCategory: [],
    productPackageNetworkSwitchType: [],
    productDiscountValue: [],
    productDiscountValueNet: [],
    productDiscountRecurrenceAmount: [],
    productDiscountRecurrenceAmountNet: [],
    productPackageName: [],
    productQuantity: [],
    transactionCouponCode: basket.voucherCode,
    quoteId: [],
    tradeInAmount: [],
    tradeInDevice: [],
    tradeInDuration: [],
    tradeInType: [],
  }

  const getBundleTitle = (packageId: BasketV2.ModelPackage['packageId']) => {
    const thisPackage = basket.packages?.find(mappedPackage => mappedPackage.packageId === packageId)
    return thisPackage?.bundle?.name?.toLowerCase()
  }

  const pushPaymentPlanCosts = (priceDetails?: BasketV2.Hardware['priceDetails']) => {
    const paymentPlan = priceDetails?.devicePaymentPlan

    const { monthlyDiscountPrice, upFrontDiscountPrice, upFrontPrice, monthlyPrice } = paymentPlan || {}

    allData.productDiscountValue.push(upFrontDiscountPrice?.gross || '0')
    allData.productDiscountValueNet.push(upFrontDiscountPrice?.net || '0')

    allData.productPrice.push(upFrontPrice?.gross || '0')
    allData.productPriceNet.push(upFrontPrice?.net || '0')

    allData.productPriceRecurrenceAmount.push(monthlyPrice?.gross || '0')
    allData.productPriceRecurrenceAmountNet.push(monthlyPrice?.net || '0')

    allData.productDiscountRecurrenceAmount.push(monthlyDiscountPrice?.gross || '0')
    allData.productDiscountRecurrenceAmountNet.push(monthlyDiscountPrice?.net || '0')
  }

  const pushCosts = (priceDetails?: BasketV2.PriceDetails) => {
    const { oneOffDiscountPrice, monthlyDiscountPrice, oneOffPrice, monthlyPrice } = priceDetails || {}

    allData.productDiscountValue.push(oneOffDiscountPrice?.gross || '0')
    allData.productDiscountValueNet.push(oneOffDiscountPrice?.net || '0')

    allData.productDiscountRecurrenceAmount.push(monthlyDiscountPrice?.gross || '0')
    allData.productDiscountRecurrenceAmountNet.push(monthlyDiscountPrice?.net || '0')

    allData.productPrice.push(oneOffPrice?.gross || '0')
    allData.productPriceNet.push(oneOffPrice?.net || '0')

    allData.productPriceRecurrenceAmount.push(monthlyPrice?.gross || '0')
    allData.productPriceRecurrenceAmountNet.push(monthlyPrice?.net || '0')
  }

  const pushAnalyticValues = (data: AnalyticValues, packageId: BasketV2.ModelPackage['packageId'], isBingoPackage?: boolean) => {
    allData.productPackageId.push(packageId || '')
    allData.productModel.push(data.model || '')
    allData.productId.push(data.skuId || '')
    allData.productQuantity.push('1')

    const dataAllowance = getAllowanceByType(data.allowances || [], 'uk data')
    const dataAllowanceValue = dataAllowance?.value?.toLowerCase() || ''
    allData.productPlanData.push(dataAllowanceValue)

    const minAllowance = getAllowanceByType(data.allowances || [], 'uk min', 'voice')
    const minAllowanceValue = minAllowance?.value?.toLowerCase() || ''
    allData.productPlanMinutes.push(minAllowanceValue)

    const textAllowance = getAllowanceByType(data.allowances || [], 'uk text')
    const textAllowanceValue = textAllowance?.value?.toLowerCase() || ''
    allData.productPlanSms.push(textAllowanceValue)

    const bundleTitle = getBundleTitle(packageId) || ''
    allData.productPackageName.push(bundleTitle)

    const productMake = data.make?.toLowerCase() || ''
    allData.productBrand.push(productMake)

    const productMemory = data.memorySize?.replace(' ', '').toLowerCase() || ''
    allData.productCapacity.push(productMemory)

    const productColour = data.deviceColor?.toLowerCase() || ''
    allData.productColour.push(productColour)

    const productName = data.name?.toLowerCase()
    const productLabel = data.label?.toLowerCase() || ''
    allData.productName.push(productName || productLabel)

    const productSwitchType = data.portability?.codeType?.toLowerCase() || ''
    allData.productPackageNetworkSwitchType.push(productSwitchType)

    if (isBingoPackage && data.priceDetails?.devicePaymentPlan) {
      pushPaymentPlanCosts(data.priceDetails)
      allData.productPackagePlanDuration.push(data.priceDetails?.devicePaymentPlan?.duration?.value || '')
    } else {
      pushCosts(data.priceDetails)
      allData.productPackagePlanDuration.push(data.commitmentPeriod?.value || '')
    }
  }

  //* Loop through all hardware and push analytic data for each */
  const getHardwareData = (hardwares: BasketV2.Hardware[], packageId: BasketV2.ModelPackage['packageId'], isBingoPackage: boolean) => {
    if (!Array.isArray(hardwares) || !hardwares.length) return
    hardwares.forEach(hardware => {
      allData.productCategory.push(hardware.productClass?.toLowerCase() || '')
      pushAnalyticValues(hardware, packageId, isBingoPackage)
    })
  }

  //* Loop through all addons and push analytic data for each */
  const getAddonData = (services: BasketV2.Service[], packageId: BasketV2.ModelPackage['packageId'], skuId: string) => {
    if (!Array.isArray(services) || !services.length) return
    services.forEach(service => {
      //* if specific addon is supplied we only want to return the data for that specific addon */
      if (!skuId || skuId === service.skuId) {
        allData.productCategory.push(service.productClass?.toLowerCase() || '')
        pushAnalyticValues(service, packageId)
      }
    })
  }

  //* Loop through all addons and push analytic data for each */
  const getAccessoryAddonData = (hardwares: BasketV2.Hardware[], packageId: BasketV2.ModelPackage['packageId'], skuId: string) => {
    if (!Array.isArray(hardwares) || !hardwares.length) {
      return
    }

    hardwares.forEach(hardware => {
      //* if specific addon is supplied we only want to return the data for that specific addon */
      if (!skuId || skuId === hardware.skuId) {
        allData.productCategory.push(hardware.productClass?.toLowerCase() || '')
        pushAnalyticValues(hardware, packageId)
      }
    })
  }

  //* push analytic data for bundle */
  const getBundleData = (bundle: BasketV2.Bundle, packageId: BasketV2.ModelPackage['packageId']) => {
    if (!bundle) return
    allData.productCategory.push('bundle')
    pushAnalyticValues(bundle, packageId)
  }

  //* Loop through all discounts and push analytic data for each */
  const getDiscountData = (discounts: BasketV2.Discount[], packageId: BasketV2.ModelPackage['packageId']) => {
    if (!Array.isArray(discounts) || !discounts.length) return
    discounts.forEach(discount => {
      allData.productCategory.push('discount')
      pushAnalyticValues(discount, packageId)
    })
  }

  if (basket.packages) {
    basket.packages.forEach(thisPackage => {
      const isBingoPackage = isBingo(thisPackage)
      if (thisPackage.tradeInCredit && (!specificPackage || thisPackage.packageId === specificPackage)) {
        const tradeInData = getTradeInParams(thisPackage.tradeInCredit)
        allData.quoteId.push(tradeInData.quoteId || '')
        allData.tradeInType.push(tradeInData.tradeInType)
        allData.tradeInDuration.push(tradeInData.tradeInDuration || 0)
        allData.tradeInAmount.push(tradeInData.tradeInAmount || 0)
        allData.tradeInDevice.push(tradeInData.tradeInDevice || '')
      }
      if (!hasAccessoryAddon && addonId && thisPackage.services) {
        return getAddonData(thisPackage.services, thisPackage.packageId, addonId)
      }
      if (hasAccessoryAddon && addonId && thisPackage.hardwares) {
        return getAccessoryAddonData(thisPackage.hardwares, thisPackage.packageId, addonId)
      }
      if (!specificPackage || (specificPackage && specificPackage === thisPackage.packageId)) {
        if (thisPackage.hardwares) {
          getHardwareData(thisPackage.hardwares, thisPackage.packageId, isBingoPackage)
        }
        if (thisPackage.services) {
          getAddonData(thisPackage.services, thisPackage.packageId, '')
        }
        if (thisPackage.bundle) {
          getBundleData(thisPackage.bundle, thisPackage.packageId)
        }
        if (thisPackage.discounts) {
          getDiscountData(thisPackage.discounts, thisPackage.packageId)
        }
      }
    })
  }

  return allData
}

export default getProductAnalyticData
