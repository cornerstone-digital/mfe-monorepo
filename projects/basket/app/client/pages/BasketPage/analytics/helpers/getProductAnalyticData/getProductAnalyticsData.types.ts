export interface ProductAnalyticsData {
  productPackageId: string[]
  productBrand: string[]
  productCapacity: string[]
  productColour: string[]
  productId: string[]
  productModel: string[]
  productName: string[]
  productPlanData: string[]
  productPackagePlanDuration: string[]
  productPlanMinutes: string[]
  productPlanSms: string[]
  productPrice: string[]
  productPriceNet: string[]
  productPriceRecurrenceAmount: string[]
  productPriceRecurrenceAmountNet: string[]
  productCategory: string[]
  productPackageNetworkSwitchType: string[]
  productDiscountValue: string[]
  productDiscountValueNet: string[]
  productDiscountRecurrenceAmount: string[]
  productDiscountRecurrenceAmountNet: string[]
  productPackageName: string[]
  productQuantity: string[]
  transactionCouponCode: BasketV2.Basket['voucherCode']
  quoteId: string[]
  tradeInDuration: number[]
  tradeInType: string[]
  tradeInAmount: number[]
  tradeInDevice: string[]
}
