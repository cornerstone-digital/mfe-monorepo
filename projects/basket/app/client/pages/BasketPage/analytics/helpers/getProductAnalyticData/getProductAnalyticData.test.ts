import simoMockJson from '@basketMocks/basketWithSimo.mock.json'
import planBenefitsJson from '@basketMocks/basketWithBingoPlanBenefits.mock.json'
import tradeinMockJson from '@basketMocks/basketWithTradeIn.mock.json'
import TransformUtils from '@shared/helpers/transformUtils/transformUtils'

import getProductAnalyticData from './getProductAnalyticData'

const simoMock = TransformUtils.removeNulls(simoMockJson)
const bingoMock = TransformUtils.removeNulls(planBenefitsJson)
const tradeinMock = TransformUtils.removeNulls(tradeinMockJson)

describe('getProductAnalytic data', () => {
  it('returns simo product analytic data', () => {
    const data = getProductAnalyticData(simoMock)
    expect(data).toMatchSnapshot()
  })

  it('returns bingo product analytic data', () => {
    const data = getProductAnalyticData(bingoMock)
    expect(data).toMatchSnapshot()
  })

  it('returns analytic data with tradein', () => {
    const data = getProductAnalyticData(tradeinMock) as any
    expect(data.tradeInAmount).toEqual([251.99, 10.41])
    expect(data.tradeInDevice).toEqual(['Apple iPhone XS 64GB', 'Apple iPhone XS 64GB'])
    expect(data.tradeInDuration).toEqual([1, 24])
    expect(data.tradeInType).toEqual(['bacs', 'monthly credit'])
    expect(data.quoteId).toEqual(['1111111111', '222222222'])
  })
})
