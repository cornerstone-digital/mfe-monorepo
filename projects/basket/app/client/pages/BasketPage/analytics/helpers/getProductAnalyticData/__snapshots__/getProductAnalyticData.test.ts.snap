// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getProductAnalytic data returns bingo product analytic data 1`] = `
Object {
  "productBrand": Array [
    "",
    "",
    "",
    "apple",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "vodafone",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "apple",
    "",
  ],
  "productCapacity": Array [
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "64gb",
    "",
  ],
  "productCategory": Array [
    "sim card",
    "bundle",
    "sim card",
    "ehandset",
    "bundle",
    "discount",
    "sim card",
    "bundle",
    "extender product",
    "sim card",
    "accessories",
    "data_device",
    "fixed broadband equipment 2019",
    "super wi-fi",
    "always connected bundle",
    "bundle",
    "discount",
    "sim card",
    "discount",
    "handset",
    "bundle",
  ],
  "productColour": Array [
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "gold",
    "",
  ],
  "productDiscountRecurrenceAmount": Array [
    "0",
    "11",
    "0",
    "0",
    "55",
    "0",
    "0",
    "26",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "42",
    "0",
    "0",
    "0",
    "0",
    "27",
  ],
  "productDiscountRecurrenceAmountNet": Array [
    "0",
    "9.17",
    "0",
    "0",
    "45.83",
    "0",
    "0",
    "21.67",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "35",
    "0",
    "0",
    "0",
    "0",
    "22.50",
  ],
  "productDiscountValue": Array [
    "0",
    "0",
    "0",
    "29",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "8.00",
    "0",
  ],
  "productDiscountValueNet": Array [
    "0",
    "0",
    "0",
    "24.17",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "6.33",
    "0",
  ],
  "productId": Array [
    "086532",
    "114361",
    "086532",
    "212492",
    "113362",
    "107389",
    "086532",
    "113243",
    "206929",
    "086532",
    "201438",
    "209128",
    "206680",
    "111997",
    "112785",
    "112759",
    "111116",
    "086532",
    "110135",
    "209112",
    "112681",
  ],
  "productModel": Array [
    "",
    "",
    "",
    "iphone-12",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "k5161-dongle",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "iphone-8",
    "",
  ],
  "productName": Array [
    "blank white triple format sim",
    "12mth data simo plan 4gb",
    "blank white triple format sim",
    "apple iphone 12 64gb purple",
    "24month band c 100gb red plan",
    "yours for a great price",
    "blank white triple format sim",
    "24month sim only unlimited plan",
    "super wifi booster",
    "blank white triple format sim",
    "apple tv 32gb",
    "vodafone 4g dongle",
    "vodafone wi-fi hub",
    "super wi-fi product",
    "broadband back up 50gb data bundle",
    "pro superfast 2 xtra",
    "vodafone together Â£3 discount",
    "blank white triple format sim",
    "vodafone household 15% discount",
    "new bingo 2",
    "bingo test b silver 24 month unlimited",
  ],
  "productPackageId": Array [
    "a5d2dfea-bec1-4ab8-82be-c99e3aefc2fe",
    "a5d2dfea-bec1-4ab8-82be-c99e3aefc2fe",
    "b2b5bfd3-7137-4a10-99d4-eaeb0cd96cea",
    "b2b5bfd3-7137-4a10-99d4-eaeb0cd96cea",
    "b2b5bfd3-7137-4a10-99d4-eaeb0cd96cea",
    "b2b5bfd3-7137-4a10-99d4-eaeb0cd96cea",
    "43650589-9efe-4109-ac1e-73903c49aed0",
    "43650589-9efe-4109-ac1e-73903c49aed0",
    "d6211217-262f-42ea-8f81-d9f1739cc1df",
    "d6211217-262f-42ea-8f81-d9f1739cc1df",
    "d6211217-262f-42ea-8f81-d9f1739cc1df",
    "d6211217-262f-42ea-8f81-d9f1739cc1df",
    "d6211217-262f-42ea-8f81-d9f1739cc1df",
    "d6211217-262f-42ea-8f81-d9f1739cc1df",
    "d6211217-262f-42ea-8f81-d9f1739cc1df",
    "d6211217-262f-42ea-8f81-d9f1739cc1df",
    "d6211217-262f-42ea-8f81-d9f1739cc1df",
    "be93962c-7744-48a4-b249-23b90667a9d1",
    "be93962c-7744-48a4-b249-23b90667a9d1",
    "be93962c-7744-48a4-b249-23b90667a9d1",
    "be93962c-7744-48a4-b249-23b90667a9d1",
  ],
  "productPackageName": Array [
    "12mth data simo plan 4gb",
    "12mth data simo plan 4gb",
    "24month band c 100gb red plan",
    "24month band c 100gb red plan",
    "24month band c 100gb red plan",
    "24month band c 100gb red plan",
    "24month sim only unlimited plan",
    "24month sim only unlimited plan",
    "pro superfast 2 xtra",
    "pro superfast 2 xtra",
    "pro superfast 2 xtra",
    "pro superfast 2 xtra",
    "pro superfast 2 xtra",
    "pro superfast 2 xtra",
    "pro superfast 2 xtra",
    "pro superfast 2 xtra",
    "pro superfast 2 xtra",
    "bingo test b silver 24 month unlimited",
    "bingo test b silver 24 month unlimited",
    "bingo test b silver 24 month unlimited",
    "bingo test b silver 24 month unlimited",
  ],
  "productPackageNetworkSwitchType": Array [
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
  ],
  "productPackagePlanDuration": Array [
    "",
    "12",
    "",
    "",
    "24",
    "",
    "",
    "24",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "24",
    "",
    "",
    "",
    "24",
    "24",
  ],
  "productPlanData": Array [
    "",
    "4",
    "",
    "",
    "200",
    "",
    "",
    "unlimited",
    "",
    "",
    "",
    "",
    "",
    "",
    "50",
    "",
    "",
    "",
    "",
    "",
    "unlimited",
  ],
  "productPlanMinutes": Array [
    "",
    "",
    "",
    "",
    "unlimited",
    "",
    "",
    "unlimited",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "unlimited",
  ],
  "productPlanSms": Array [
    "",
    "",
    "",
    "",
    "unlimited",
    "",
    "",
    "unlimited",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "",
    "unlimited",
  ],
  "productPrice": Array [
    "0",
    "0",
    "0",
    "29",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "10.00",
    "0",
  ],
  "productPriceNet": Array [
    "0",
    "0",
    "0",
    "24.17",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "8.33",
    "0",
  ],
  "productPriceRecurrenceAmount": Array [
    "0",
    "11",
    "0",
    "0",
    "59",
    "0",
    "0",
    "26",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "45",
    "0",
    "0",
    "0",
    "1.67",
    "27",
  ],
  "productPriceRecurrenceAmountNet": Array [
    "0",
    "9.17",
    "0",
    "0",
    "49.17",
    "0",
    "0",
    "21.67",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "0",
    "37.50",
    "0",
    "0",
    "0",
    "1.39",
    "22.50",
  ],
  "productQuantity": Array [
    "1",
    "1",
    "1",
    "1",
    "1",
    "1",
    "1",
    "1",
    "1",
    "1",
    "1",
    "1",
    "1",
    "1",
    "1",
    "1",
    "1",
    "1",
    "1",
    "1",
    "1",
  ],
  "quoteId": Array [],
  "tradeInAmount": Array [],
  "tradeInDevice": Array [],
  "tradeInDuration": Array [],
  "tradeInType": Array [],
  "transactionCouponCode": undefined,
}
`;

exports[`getProductAnalytic data returns simo product analytic data 1`] = `
Object {
  "productBrand": Array [
    "",
    "",
    "",
  ],
  "productCapacity": Array [
    "",
    "",
    "",
  ],
  "productCategory": Array [
    "sim card",
    "bundle",
    "discount",
  ],
  "productColour": Array [
    "",
    "",
    "",
  ],
  "productDiscountRecurrenceAmount": Array [
    "0",
    "15",
    "0",
  ],
  "productDiscountRecurrenceAmountNet": Array [
    "0",
    "12.50",
    "0",
  ],
  "productDiscountValue": Array [
    "0",
    "0",
    "0",
  ],
  "productDiscountValueNet": Array [
    "0",
    "0",
    "0",
  ],
  "productId": Array [
    "086532",
    "112168",
    "109989",
  ],
  "productModel": Array [
    "",
    "",
    "",
  ],
  "productName": Array [
    "blank white triple format sim",
    "24m sim only unlimited max plan",
    "simo limited time discount for 6 months ",
  ],
  "productPackageId": Array [
    "3023bbbd-5552-4c6f-8283-0a5b267a3d3c",
    "3023bbbd-5552-4c6f-8283-0a5b267a3d3c",
    "3023bbbd-5552-4c6f-8283-0a5b267a3d3c",
  ],
  "productPackageName": Array [
    "24m sim only unlimited max plan",
    "24m sim only unlimited max plan",
    "24m sim only unlimited max plan",
  ],
  "productPackageNetworkSwitchType": Array [
    "",
    "",
    "",
  ],
  "productPackagePlanDuration": Array [
    "",
    "24",
    "",
  ],
  "productPlanData": Array [
    "",
    "unlimited",
    "",
  ],
  "productPlanMinutes": Array [
    "",
    "unlimited",
    "",
  ],
  "productPlanSms": Array [
    "",
    "unlimited",
    "",
  ],
  "productPrice": Array [
    "0",
    "0",
    "0",
  ],
  "productPriceNet": Array [
    "0",
    "0",
    "0",
  ],
  "productPriceRecurrenceAmount": Array [
    "0",
    "30",
    "0",
  ],
  "productPriceRecurrenceAmountNet": Array [
    "0",
    "25",
    "0",
  ],
  "productQuantity": Array [
    "1",
    "1",
    "1",
  ],
  "quoteId": Array [],
  "tradeInAmount": Array [],
  "tradeInDevice": Array [],
  "tradeInDuration": Array [],
  "tradeInType": Array [],
  "transactionCouponCode": undefined,
}
`;
